File: crates/klyron_sqlite/bindings/cpp/builder.cpp
#include "builder.hpp"

namespace klyron {

class Klyron::SqliteBuilder::Impl {
public:
  Klyron::SqliteConfig config;
  bool verbose = false;
};

Klyron::SqliteBuilder::Klyron::SqliteBuilder()
  : impl_(std::make_unique<Impl>()) {
}

Klyron::SqliteBuilder& Klyron::SqliteBuilder::withConfig(const Klyron::SqliteConfig& config) {
  impl_->config = config;
  return *this;
}

Klyron::SqliteBuilder& Klyron::SqliteBuilder::verbose(bool v) {
  impl_->verbose = v;
  return *this;
}

Klyron::SqliteBuilder::Klyron::SqliteInstance Klyron::SqliteBuilder::build() {
  Klyron::SqliteInstance inst;
  inst.config = impl_->config;
  inst.verbose = impl_->verbose;
  return inst;
}

} // namespace klyron

File: crates/klyron_sqlite/bindings/rust/builder.rs
//! Builder pattern for klyron_sqlite

use crate::types::Klyron::SqliteConfig;

#[derive(Debug, Default)]
pub struct Klyron::SqliteBuilder {
    config: Option<Klyron::SqliteConfig>,
    verbose: bool,
}

impl Klyron::SqliteBuilder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_config(mut self, config: Klyron::SqliteConfig) -> Self {
        self.config = Some(config);
        self
    }

    pub fn verbose(mut self, verbose: bool) -> Self {
        self.verbose = verbose;
        self
    }

    pub fn build(self) -> anyhow::Result<Klyron::SqliteInstance> {
        let config = self.config.unwrap_or_default();
        Ok(Klyron::SqliteInstance { config, verbose: self.verbose })
    }
}

#[derive(Debug)]
pub struct Klyron::SqliteInstance {
    pub config: Klyron::SqliteConfig,
    pub verbose: bool,
}
